server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -321,7 +321,7 @@ private static void checkNestedScopeCompatibility(String source, String target)
     public abstract Builder getMergeBuilder();
 
     @Override
-    public final FieldMapper merge(Mapper mergeWith) {
+    public final FieldMapper merge(Mapper mergeWith, MapperBuilderContext mapperBuilderContext) {
```

Comment:
I went for propagating the MapperBuilderContext instead of adding a String path argument, because in some cases we already have a mapper builder context instance that we can just propagate, and in general it seems like a good idea to start merging at the root level with a ROOT mapper builder context and then build a child context for each level, in the caller of the merge methods.